debian/readsb.1
.TH "readsb" 1 "12 December 2019" "" ""
.RS
.SH NAME
readsb \- Mode-S aircraft data processor for various receiver hardware.
.SH VERSION
v3.9.0
.SH DESCRIPTION
\fBreadsb Mode-S Receiver
.SH BUILD OPTIONS
.IP \(bu 3
ENABLE_RTLSDR
.IP \(bu 3
ENABLE_BLADERF
.SH USAGE
\fBreadsb [OPTION\.\.\.]
.SH OPTIONS
.TP
.B
\fB--dcfilter\fP
Apply a 1Hz DC filter to input data (requires more
CPU)
.TP
.B
\fB--enable-biastee\fP
Enable bias tee on supporting interfaces (default: disabled)
.TP
.B
\fB--debug\fP=<flags>
Debug mode (verbose), see flags below
.TP
.B
\fB--device-type\fP=<type>
Select SDR type
.TP
.B
\fB--fix\fP
Enable CRC single-bit error correction (default)
.TP
.B
\fB--forward-mlat\fP
Allow forwarding of received mlat results to
output ports
.TP
.B
\fB--freq\fP=<hz>
Set frequency (default: 1090 MHz)
.TP
.B
\fB--gain\fP=<db>
Set gain (default: max gain. Use \fB-10\fP for
auto-gain)
.TP
.B
\fB--gnss\fP
Show altitudes as GNSS when available
.TP
.B
\fB--interactive\fP
Interactive mode refreshing data on screen.
Implies \fB--throttle\fP
.TP
.B
\fB--interactive-ttl\fP=<sec>
Remove from list if idle for <sec> (default: 60)
.TP
.B
\fB--json-location-accuracy\fP=<n>
Accuracy of receiver location in json metadata:
.RS
.IP 0
no location
.IP 1
approximate
.IP 2
exact
.RE
.RE
.TP
.B
\fB--lat\fP=<lat>
Reference/receiver surface latitude
.TP
.B
\fB--lon\fP=<lon>
Reference/receiver surface longitude
.TP
.B
\fB--max-range\fP=<dist>
Absolute maximum range for position decoding (in
nm, default: 300)
.TP
.B
\fB--metric\fP
Use metric units
.TP
.B
\fB--mlat\fP
Display raw messages in Beast ASCII mode
.TP
.B
\fB--modeac\fP
Enable decoding of SSR Modes 3/A & 3/C
.TP
.B
\fB--no-crc-check\fP
Disable messages with invalid CRC (discouraged)
.TP
.B
\fB--no-fix\fP
Disable CRC single-bit error correction
.TP
.B
\fB--no-interactive\fP
Disable interactive mode, print to stdout
.TP
.B
\fB--no-modeac-auto\fP
Don't enable Mode A/C if requested by a Beast
connection
.TP
.B
\fB--onlyaddr\fP
Show only ICAO addresses
.TP
.B
\fB--quiet\fP
Disable output. Use for daemon applications
.TP
.B
\fB--raw\fP
Show only messages hex values
.TP
.B
\fB--show-only\fP=<addr>
Show only messages by given ICAO on stdout
.TP
.B
\fB--snip\fP=<level>
Strip IQ file removing samples < level
.TP
.B
\fB--stats\fP
With \fB--ifile\fP print stats at exit. No other output
.TP
.B
\fB--stats-every\fP=<sec>
Show and reset stats every <sec> seconds
.TP
.B
\fB--stats-range\fP
Collect/show range histogram
.TP
.B
\fB--write-json\fP=<dir>
Periodically write json output to <dir> (for
external webserver)
.TP
.B
\fB--write-json-every\fP=<t>
Write json output every t seconds (default 1)
.SS  NETWORK OPTIONS
.TP
.B
\fB--net\fP
Enable networking
.TP
.B
\fB--net-bi-port\fP=<ports>
TCP Beast input listen ports (default:
30004,30104)
.TP
.B
\fB--net-vrs-port\fP=<ports>
TCP VRS json output listen ports (default: 0)
.TP
.B
\fB--net-beast-reduce-out-port\fP=<ports>
TCP BeastReduce output listen ports (default: 0)
.TP
.B
\fB--net-beast-reduce-interval\fP=<seconds>
BeastReduce position update interval, longer means less data
(default: 0.125, valid range: 0.000 - 14.999)
.TP
.B
\fB--net-bind-address\fP=<ip>
IP address to bind to (default: Any; Use 127.0.0.1 for private)
.TP
.B
\fB--net-bo-port\fP=<ports>
TCP Beast output listen ports (default: 30005)
.TP
.B
\fB--net-buffer\fP=<n>
TCP buffer size 64Kb * (2^n) (default: n=0, 64Kb)
.TP
.B
\fB--net-heartbeat\fP=<rate>
TCP heartbeat rate in seconds (default: 60 sec, 0 to disable)
.TP
.B
\fB--net-only\fP
Enable just networking, no RTL device or file used
.TP
.B
\fB--net-connector\fP=<ip,port,protocol>
Establish connection, can be specified multiple times (example: 127.0.0.1,23004,beast_out) Protocols: beast_out, beast_in, raw_out, raw_in, sbs_out, vrs_out.
.TP
.B
\fB--net-connector-delay\fP=<seconds>
Outbound re-connection delay (default: 30)
.TP
.B
\fB--net-ri-port\fP=<ports>
TCP raw input listen ports  (default: 30001)
.TP
.B
\fB--net-ro-interval\fP=<rate>
TCP output memory flush rate in seconds
(default: 0)
.TP
.B
\fB--net-ro-port\fP=<ports>
TCP raw output listen ports (default: 30002)
.TP
.B
\fB--net-ro-size\fP=<size>
TCP output minimum size (default: 0)
.TP
.B
\fB--net-sbs-port\fP=<ports>
TCP BaseStation output listen ports (default: 30003)
.RE
.TP
.B
\fB--net-verbatim\fP
Forward messages unchanged
.SS  RTLSDR OPTIONS
.I
use with \fB--device-type\fP rtlsdr
.TP
.B
\fB--device\fP=<index|serial>
Select device by index or serial number
.TP
.B
\fB--enable-agc\fP
Enable digital AGC (not tuner AGC!)
.TP
.B
\fB--ppm\fP=<correction>
Set oscillator frequency correction in PPM
.SS  BLADERF OPTIONS
.I
use with \fB--device-type\fP bladerf
.TP
.B
\fB--bladerf-bandwidth\fP=<hz>
Set LPF bandwidth ('bypass' to bypass the
LPF)
.TP
.B
\fB--bladerf-decimation\fP=<N>
Assume FPGA decimates by a factor of N
.TP
.B
\fB--bladerf-fpga\fP=<path>
Use alternative FPGA bitstream ('' to disable FPGA load)
.TP
.B
\fB--device\fP=<ident>
Select device by bladeRF 'device identifier'
.SS  MODES BEAST OPTIONS
.I
use with \fB--device-type\fP modesbeast
.PP
Beast binary protocol and hardware handshake are always enabled.
.TP
.B
\fB--beast-crc-off\fP
Turn OFF CRC checking
.TP
.B
\fB--beast-df045-on\fP
Turn ON DF0/4/5 filter
.TP
.B
\fB--beast-df1117-on\fP
Turn ON DF11/17-only filter
.TP
.B
\fB--beast-fec-off\fP
Turn OFF forward error correction
.TP
.B
\fB--beast-mlat-off\fP
Turn OFF MLAT time stamps
.TP
.B
\fB--beast-modeac\fP
Turn ON mode A/C
.TP
.B
\fB--beast-serial\fP=<path>
Path to Beast serial device (default /dev/ttyUSB0)
.SS  GNS HULC OPTIONS
.I
use with \fB--device-type\fP gnshulc
.PP
Beast binary protocol and hardware handshake are always enabled.
.TP
.B
\fB--beast-serial\fP=<path>
Path to GNS HULC serial device (default /dev/ttyUSB0)
.SS  ADALM-Pluto OPTIONS
.I
use with \fB--device-type\fP plutosdr
.PP
Default device expected at hostname 'pluto.local'.
.TP
.B
\fB--pluto-uri\fP=<USB URI>
Create USB context from URI (eg. usb:1.2.5)
.TP
.B
\fB--pluto-network\fP=<hostname or IP>
Create network context from hostname or IP (default pluto.local)
.SS  IFILE OPTIONS
.I
use with \fB--ifile\fP
.TP
.B
\fB--ifile\fP=<path>
Read samples from given file ('-' for stdin)
.TP
.B
\fB--iformat\fP=<type>
Set sample format (UC8, SC16, SC16Q11)
.TP
.B
\fB--throttle\fP
Process samples at the original capture speed
.SS  HELP OPTIONS
.TP
.B
-?, \fB--help\fP
Give this help list
.TP
.B
\fB--usage\fP
Give a short usage message
.TP
.B
\fB-V\fP, \fB--version\fP
Print program version
.SH DEBUG MODE FLAGS
.TP
.B
d
Log frames decoded with errors
.TP
.B
D
Log frames decoded with zero errors
.TP
.B
c
Log frames with bad CRC
.TP
.B
C
Log frames with good CRC
.TP
.B
p
Log frames with bad preamble
.TP
.B
n
Log network debugging info
.TP
.B
j
Log frames to frames.js, loadable by debug.html
.PP
.SH CREDITS
Matthias Wirth aka wiedehopf
Taner Halicioglu aka tanerH
.SH BUGS
Report bugs to Michael Wolf <michael@mictronics.de>.
.SH AUTHOR
Michael Wolf


debian/changelog
readsb (3.9.0) stable; urgency=medium

  * Last and final release of this readsb version.
  * Support GNS HULC antenna.
  * Interrupt driven serial device handling removed.
  * Clean up linkage of struct Modes. (via dump1090-fa)
  * Build with -fno-common to match GCC 10 behaviour on older compiler versions. (via dump1090-fa)
  * Use precise location on local map.
  * Contribution wiedehopf:
    * Improve altitude filter.
    * Show zero timestamps using viewadsb.

 -- Michael Wolf <michael@mictronics.de>  Sat, 01 Aug 2020 10:20:10 +0200

readsb (3.8.3) stable; urgency=medium

  * Add marker types for ground vehicles and towers.
  * Fix bug in handling bladeRF command line options.
  * Contribution wiedehopf:
    * Improvement of SBS decoding
    * Connection timed out: remove misleading explanation
    * keepalive: set and check for error in net_io.c instead of blindly doing it in anet.c
    * Add custom keepalive settings that will time out a broken connection after 30 seconds.
    * Move from global client list to one client list per service.
    * Fix anet open_fd limiting to include all used sockets.
    * Avoid excessive iteration over Modes.clients.
    * Move network cleanup to net_io.c and use closeClient.
    * Perform some network task less often.
    * Improved accept handling.

 -- Michael Wolf <michael@mictronics.de>  Sun, 03 May 2020 10:30:05 +0200

readsb (3.8.2) stable; urgency=medium

  * Webapp: History loading and trace collection moved to background worker.
  * Fix hidding SkyVector layers when no API key is present.
  * Fix latest stats, was always zero, is now momentary statistic.
  * Fix bug in overlay layer generation when language is not English.
  * Add DC-10 aircraft marker.
  * Disable source maps for third-party scripts.
  * Fix sporatic buffer overflow in JSON stats generation.
  * Zero initalize new network clients.
  * Prevent segmentation fault on invalid protocol name.

 -- Michael Wolf <michael@mictronics.de>  Sun, 01 Mar 2020 10:02:17 +0100

readsb (3.8.1) stable; urgency=medium

  * Load default web application settings from JSON file.
  * Introduce a dark theme.
  * Via dump1090:
    * Treat ADS-R as a separate source type.
    * Track ADS-B version separately for direct ADS-B, TIS-B, ADS-R.
    * Compute ADS-B v0 NACp/SIL from position message type
    * Correct block duration with PlutoSDR, fix issue #17.
  * Contribution by wiedehopf and tanerH:
    * Increase network buffers.
    * Removed redundant history code.
    * Filtering for implausible altitudes.
    * Use multible linked lists for aircraft list.
    * Add max distance to stats.json.
    * Network improvements and fixes.
    * Add VRS JSON network writer.
    * Fixed push client.
    * More statistics for beast network input.
    * Improved VRS JSON generation.
    * Implementation of non-blocking TCP connection.
    * Improved handling of multible address records.
    * Removed legacy push server commands. Replaced by --net-connector.
    * BeastReduce output - Reduced message rate for non-physical aircraft data.
    * Aynchronous name resolution based on pthread.
    * Fix network connection in viewadsb.

 -- Michael Wolf <michael@mictronics.de>  Mon, 06 Jan 2020 09:43:24 +0200

readsb (3.8.0) stable; urgency=medium

  * Multi language support in web application.
  * Uses I18next for internationalization.
  * Web application reworked and ported to Typescript.
  * Moved to Leaflet map library.
  * Replace obsolete usleep in POSIX 2008 with nanosleep.
  * Improvements to position filtering and relative decoding.
  * Position decoding bugfixes.
  * When generating aircraft.json, leave space for the final line.
  * Forward alert and spi bits to web application.
  * Added support for GNS5894 receiver hardware.

 -- Michael Wolf <michael@mictronics.de>  Tue, 03 Dec 2019 19:11:17 +0200

readsb (3.7.3) stable; urgency=medium

  * First stable release under new package name readsb.
  * Improve package installation and removal.
  * Fork detached from dump1090-fa into standalone project and renamed to readsb.
  * Accept profiles to build package with indivitual or no receiver library dependencies.

 -- Michael Wolf <michael@mictronics.de>  Tue, 03 Sep 2019 18:52:56 +0200

readsb (3.7.2) stable; urgency=medium

  * Added configurable re-connection delay for push server support.
  * Add bladeRF v2.0 Micro support (credits @kazazes)

 -- Michael Wolf <michael@mictronics.de>  Fri, 26 Jul 2019 18:47:32 +0200

readsb (3.7.1) stable; urgency=medium

  * Update aircraft database.
  * Fix web app to make links and button in selected plane dialog accessible.
  * Removed obsolete code and styles in web app.
  * jQuery touchpunch updated to bugfixed fork.
  * Extended aircraft species filter and added engine type filter.
  * Add OpenFlightMap layer.
  * Update to OpenLayers v5.3.3 and fix code compatibility.
  * Fix array out of bounds with gcc-8.
  * Fix nav_altitude and nav_heading description in README-json.
  * Bump TSV_VERSION for TAH sense fix.
  * Fix reversed sense of Track Angle / Heading bit in surface opstatus messages.
  * Some corrections to JSON doc updates. (upstream /flightaware/dump1090 commit)
  * Bump TSV_VERSION to reflect extensive comm-b changes. (upstream /flightaware/dump1090 commit)
  * Improved comm-b decoding and testing. (several upstream /flightaware/dump1090 commits)
  * Show configurable PageName also in right hand pane info block.
  * Use RTLSDR as default device type.
  * Add configuration hint for clean installation.
  * Add support and build option for bias tee.
  * Keep backward compatible with pre-2.x bladeRF versions.
  * Fix an error in the bit range used to decode CF in DF18 messages.
  * Update README-json to include and explain new fields.
  * Increase maximum TSV packet size
  * Fix possible missing \n termination of TSV messages
  * Fix column index to hide distance when site is not shown.
  * Fix getting LAT/LON from piaware .env file.

webapp/src/script/readsb/readsb.ts
// Part of readsb, a Mode-S/ADSB/TIS message decoder.
//
// readsb.ts: Main class for readsb web application.
//
// Copyright (c) 2020 Michael Wolf <michael@mictronics.de>
//
// This file is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// any later version.
//
// This file is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

namespace READSB {
    export class Main {
        /**
         * Initialize web application.
         */
        public static Initialize() {
            Registration.Init();
            Body.Init();
            Body.InitEventHandler();
            Body.InitDropDown();
            Input.InitializeCheckboxes();
            Input.SetSiteCirclesDistancesInput();
            Filter.Initialize();
            this.SetLanguage(AppSettings.AppLanguage);

            // Get the aircraft list row template from HTML.
            AircraftCollection.RowTemplate = Body.GetAircraftListRowTemplate();

            // Maybe hide flag info
            Body.ShowFlags(AppSettings.ShowFlags);

            // Hide some aircraft list columns when table is not expanded.
            Body.AircraftListSetColumnVisibility(false);

            // Sort aircraft list first time depending on site status.
            if (typeof AppSettings.SiteLat === "number" && typeof AppSettings.SiteLon === "number") {
                Input.SetSiteCoordinates();
                AircraftCollection.SortByDistance();
            } else {
                AircraftCollection.RowTemplate.cells[10].classList.add("hidden"); // hide distance column
                Body.AircraftListShowColumn("#aircraftListDistance", false); // hide distance header
                AircraftCollection.SortByAltitude();
            }

            // Get receiver metadata, reconfigure using it, then continue
            // with initialization
            fetch("data/receiver.json", {
                cache: "no-cache",
                method: "GET",
                mode: "cors",
            })
                .then((res: Response) => {
                    if (res.status >= 200 && res.status < 300) {
                        return Promise.resolve(res);
                    } else {
                        return Promise.reject(new Error(res.statusText));
                    }
                })
                .then((res: Response) => {
                    return res.json();
                })
                .then((data: IReceiverJson) => {
                    if (typeof data.lat !== "undefined") {
                        AppSettings.SiteLat = data.lat;
                        AppSettings.SiteLon = data.lon;
                        AppSettings.CenterLat = data.lat;
                        AppSettings.CenterLon = data.lon;
                    }

                    this.readsbVersion = data.version;

                    this.dataRefreshInterval = data.refresh;
                    // Start loading history
                    AircraftCollection.Init(data.history);

                    // RestoreSessionFilters();

                    this.RefreshAircraftListTable();
                    Body.RefreshInfoBlock(this.readsbVersion, this.GetMessageRate());
                    Body.RefreshSelectedAircraft();
                    AircraftCollection.Clean();
                    console.info("Completing init");

                    // Setup our timer to poll from the server.
                    window.setInterval(Main.FetchData.bind(Main), Main.DataRefreshInterval);
                    window.setInterval(AircraftCollection.Clean.bind(AircraftCollection), 60000);

                    // And kick off one refresh immediately.
                    Main.FetchData();
                });
        }

        /**
         * Set application language.
         * @param lng Language to set (ISO-639-1 code)
         */
        public static SetLanguage(lng: string) {
            // Make english the default language in failure cases
            if (lng === "" || lng === null || lng === undefined) {
                lng = "en";
            }

            i18next.use(i18nextXHRBackend).init({
                backend: {
                    loadPath: `./locales/${lng}.json`,
                },
                debug: false,
                fallbackLng: "en",
                lng,
            }, (err, t) => {
                const localize = LocI18next.Init(i18next);
                localize(".localized");
                Strings.OnLanguageChange();
                Body.UpdateAircraftListColumnUnits();
                // Init map when i18next is initialized to translate its strings.
                // No initialization when language changes.
                if (!LMap.Initialized) {
                    LMap.Init();
                }
            });
        }

        /**
         * Fetch data from readsb backend service.
         * Periodical called.
         */
        public static FetchData() {
            if (this.fetchPending) {
                // don't double up on fetches, let the last one resolve
                return;
            }

            this.fetchPending = true;
            fetch("data/aircraft.json", {
                cache: "no-cache",
                method: "GET",
                mode: "cors",
            })
                .then((res: Response) => {
                    if (res.status >= 200 && res.status < 300) {
                        return Promise.resolve(res);
                    } else {
                        return Promise.reject(new Error(res.statusText));
                    }
                })
                .then((res: Response) => {
                    return res.json();
                })
                .then((data: IAircraftData) => {
                    const now = data.now;
                    // Detect stats reset
                    if (this.messageCountHistory.length > 0 && this.messageCountHistory[this.messageCountHistory.length - 1].messages > data.messages) {
                        this.messageCountHistory = [{
                            messages: 0,
                            time: this.messageCountHistory[this.messageCountHistory.length - 1].time,
                        }];
                    }

                    // Note the message count in the history
                    this.messageCountHistory.push({ time: now, messages: data.messages });
                    // and clean up any old values
                    if ((now - this.messageCountHistory[0].time) > 30) {
                        this.messageCountHistory.shift();
                    }

                    // Update aircraft data, timestamps, visibility, history track for all aircrafts.
                    AircraftCollection.Update(data, now, this.lastReceiverTimestamp);

                    this.RefreshAircraftListTable();
                    Body.RefreshInfoBlock(this.readsbVersion, this.GetMessageRate());
                    Body.RefreshSelectedAircraft();

                    // Check for stale receiver data
                    if (this.lastReceiverTimestamp === now) {
                        this.staleReceiverCount++;
                        if (this.staleReceiverCount > 5) {
                            Body.UpdateErrorToast(i18next.t("error.dataTimeOut"), true);
                        }
                    } else {
                        this.staleReceiverCount = 0;
                        this.lastReceiverTimestamp = now;
                        Body.UpdateErrorToast("", false);
                    }
                    this.fetchPending = false;
                })
                .catch((error) => {
                    this.fetchPending = false;
                    Body.UpdateErrorToast(i18next.t("error.fetchingData", { msg: error }), true);
                    console.error(error);
                });
        }

        private static dataRefreshInterval: number = 0;
        static get DataRefreshInterval(): number {
            return this.dataRefreshInterval;
        }
        private static readsbVersion: string;
        private static fetchPending: boolean = false;
        private static staleReceiverCount: number = 0;
        private static lastReceiverTimestamp: number = 0;
        private static messageCountHistory: IMessageCountHistory[] = [];

        /**
         * Refreshes the aircraft list table in GUI.
         */
        private static RefreshAircraftListTable() {
            AircraftCollection.TrackedAircrafts = 0;
            AircraftCollection.TrackedAircraftPositions = 0;
            AircraftCollection.TrackedAircraftUnknown = 0;
            AircraftCollection.TrackedHistorySize = 0;
            AircraftCollection.Refresh();
            AircraftCollection.ResortList();
        }

        private static GetMessageRate(): number {
            let messageRate: number = null;
            if (this.messageCountHistory.length > 1) {
                const messageTimeDelta = this.messageCountHistory[this.messageCountHistory.length - 1].time - this.messageCountHistory[0].time;
                const messageCountDelta = this.messageCountHistory[this.messageCountHistory.length - 1].messages - this.messageCountHistory[0].messages;
                if (messageTimeDelta > 0) {
                    messageRate = messageCountDelta / messageTimeDelta;
                }
            } else {
                messageRate = null;
            }
            return messageRate;
        }
    }
}
